kythe/go/serving/pipeline/beam.go

```diff
@@ -503,3 +513,118 @@ func refKind(r *ppb.Reference) string {
 	}
 	return r.GetGenericKind()
 }
+
+// Edges returns a Kythe edges table derived from the Kythe input graph.  The beam.PCollections have
+// elements of type KV<string, *srvpb.PagedEdgeSet> and KV<string, *srvpb.EdgePage>, respectively.
+func (k *KytheBeam) Edges() (beam.PCollection, beam.PCollection) {
+	s := k.s.Scope("Edges")
+
+	nodes := beam.ParDo(s, moveSourceToKey, k.nodes)
+	edges := beam.ParDo(s, reverseEdge, beam.CoGroupByKey(s, nodes, beam.ParDo(s, nodeToEdges, k.nodes)))
+	rev := beam.ParDo(s, nodeToReverseEdges, k.nodes)
+
+	return beam.ParDo2(s, groupEdges, beam.CoGroupByKey(s, nodes, edges, rev))
+}
+
+// nodeToReverseEdges emits an *ppb.Edge with its SourceNode populated for each of n's edges.  The
+// key for each *ppb.Edge is its Target VName.
+func nodeToReverseEdges(n *ppb.Node, emit func(*spb.VName, *ppb.Edge)) {
+	for _, e := range n.Edge {
+		emit(e.Target, &ppb.Edge{
```

Comment:
I'm not convinced.  This would be the equivalent:

```
edge := *e // shallow copy (or create a newEdge func)
edge.Source = nil
edge.SourceNode = n
emit(e.Target, &edge)
```

I find that the existing code is more readable:

```
emit(e.Target, &ppb.Edge{
  SourceNode: n,
  Target:     e.Target,
  Kind:       e.Kind,
  Ordinal:    e.Ordinal,
})
```

Reply:
Ah, hadn't noticed that there's a `Source` field that needs to be nil-ed out.  That could be done inside the `newEdge()` function, but I understand why you might prefer the status quo.